src/llm/exec.rs

```diff
@@ -0,0 +1,166 @@
+    /// such as schema, equivalence properties, ordering, partitioning, etc.
+    fn compute_properties(
+        input: &Arc<dyn ExecutionPlan>,
+        schema: SchemaRef,
+    ) -> Result<PlanProperties> {
+        let eq_properties = EquivalenceProperties::new(schema);
+
+        // TODO: This is a dummy partitioning. We need to figure out the actual partitioning.
+        let output_partitioning = Partitioning::RoundRobinBatch(1);
+
+        Ok(PlanProperties::new(
+            eq_properties,
+            output_partitioning,
+            input.pipeline_behavior(),
+            input.boundedness(),
+        ))
+    }
+}
+
+impl DisplayAs for AsyncFuncExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                let expr: Vec<String> = self
+                    .async_exprs
+                    .iter()
+                    .map(|async_expr| async_expr.to_string())
+                    .collect();
+
+                write!(f, "AsyncFuncExec: async_expr=[{}]", expr.join(", "))
+            }
+        }
+    }
+}
+
+impl ExecutionPlan for AsyncFuncExec {
+    fn name(&self) -> &str {
+        "async_func"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn properties(&self) -> &PlanProperties {
+        &self.cache
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        vec![&self.input]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(AsyncFuncExec::new(
+            self.async_exprs.clone(),
+            Arc::clone(&self.input),
+        )))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        trace!(
+            "Start AsyncFuncExpr::execute for partition {} of context session_id {} and task_id {:?}",
+            partition,
+            context.session_id(),
+            context.task_id()
+        );
+        // TODO figure out how to record metrics
+
+        // first execute the input stream
+        let input_stream = self.input.execute(partition, context.clone())?;
+
+        // now, for each record batch, evaluate the async expressions and add the columns to the result
+        let async_exprs_captured = Arc::new(self.async_exprs.clone());
+        let schema_captured = self.schema();
+
+        let stream_with_async_functions = input_stream.then(move |batch| {
+            // need to clone *again* to capture the async_exprs and schema in the
+            // stream and satisfy lifetime requirements.
+            let async_exprs_captured = Arc::clone(&async_exprs_captured);
+            let schema_captured = schema_captured.clone();
+
+            async move {
+                let batch = batch?;
```

Comment:
minor, would moving this invocation of the `?` operator save a task in case of an error?

Reply:
yes, you are right -- that would be an improvement 👍